File: Python/is_subsequence.py
# https://leetcode.com/problems/is-subsequence/description/
from utils.examination import Examiner, Testcase


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if len(s) > len(t):
            return False
        s_i: int = 0
        t_i: int = 0
        while s_i < len(s) and t_i < len(t):
            if s[s_i] == t[t_i]:
                s_i += 1
            t_i += 1
        return s_i >= len(s)

def examine()->None:
    Examiner(Solution(), 'isSubsequence').run((
        Testcase(('', ''), True),
        Testcase(('A', 'c'), False),
        Testcase(('abc', 'ahbgdc'), True),
        Testcase(('axc', 'ahbgdc'), False),
    ))

File: TypeScript/solutions/is-subsequence.ts
// https://leetcode.com/problems/is-subsequence/description/
import { Examiner, Testcase } from '../utils';


function isSubsequence(s: string, t: string): boolean {
    if (s.length > t.length) return false;
    let sI = 0, tI = 0;
    while (sI < s.length && tI < t.length) {
        if (s[sI] === t[tI]) sI++;
        tI++;
    }
    return sI >= s.length;
}

export function examine(): void {
    new Examiner(isSubsequence).run([
        new Testcase(['', ''], true),
        new Testcase(['A', 'c'], false),
        new Testcase(['abc', 'ahbgdc'], true),
        new Testcase(['axc', 'ahbgdc'], false),
    ]);
}
